### PR TITLE
Fixer for https://github.com/dotnet/runtime/issues/33785

### PR DESCRIPTION
**Pull request by pgovind**
Offers a code-fix for the following cases when they appear inside a binary expression such as `if (string.IndexOf() == -1)`:

``` csharp
String.IndexOf(char) -> string.Contains(char)
String.IndexOf(string) -> string.Contains(string, StringComparison.CurrentCulture)
String.IndexOf(string/char, StringComparison.Ordinal) -> string.Contains(string/char)
String.IndexOf(string/char, NON StringComparison.Ordinal) -> string.Contains(string/char, NON StringComparison.Ordinal)
```

The analyzer also tags the following case, but @mavasani recommended against offering a code-fix here(since it involves more data flow analysis than roslyn provides currently). Note: The analyzer will still fire:
``` csharp
int index = str.IndexOf("AString");
if (index == -1)
```

### CODE REVIEW

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferStringContainsOverIndexOf.Fixer.cs [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+                var compilation = semanticModel.Compilation;
+                if (!compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemString, out INamedTypeSymbol? stringType) ||
+                    !compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemChar, out INamedTypeSymbol? charType) ||
+                    !compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemStringComparison, out INamedTypeSymbol? stringComparisonType))
+                {
+                    return;
+                }
+
+                var operation = semanticModel.GetOperation(expression, cancellationToken);
+                // Not offering a code-fix for the variable declaration case
```
Comment (mavasani) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferStringContainsOverIndexOf.Fixer.cs:42]: Nit: Newline before starting comment
```suggestion

                // Not offering a code-fix for the variable declaration case
```

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferStringContainsOverIndexOf.Fixer.cs [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+        public sealed override FixAllProvider GetFixAllProvider() => WellKnownFixAllProviders.BatchFixer;
+
+        public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            Document doc = context.Document;
+            CancellationToken cancellationToken = context.CancellationToken;
+            SyntaxNode root = await doc.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            if (root.FindNode(context.Span) is SyntaxNode expression)
+            {
```
Comment (mavasani) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferStringContainsOverIndexOf.Fixer.cs:31]: Nit: Reduce nesting level for readability.
```suggestion
            if (!(root.FindNode(context.Span) is SyntaxNode expression))
            {
                return;
            }
```

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferStringContainsOverIndexOf.Fixer.cs [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+                        title: MicrosoftNetCoreAnalyzersResources.PreferStringContainsOverIndexOfTitle,
+                        createChangedDocument: c => ReplaceBinaryOperationWithContains(doc, instanceOperation.Syntax, root, invocationOperation.Arguments, binaryOperation, c),
+                        equivalenceKey: MicrosoftNetCoreAnalyzersResources.PreferStringContainsOverIndexOfMessage),
+                    context.Diagnostics);
+                return;
+
+                async Task<Document?> ReplaceBinaryOperationWithContains(Document document, SyntaxNode syntaxNode, SyntaxNode treeRoot, ImmutableArray<IArgumentOperation> indexOfMethodArguments, IBinaryOperation binaryOperation, CancellationToken cancellationToken)
+                {
+                    DocumentEditor editor = await DocumentEditor.CreateAsync(document, cancellationToken).ConfigureAwait(false);
+                    SyntaxEditor syntaxEditor = new SyntaxEditor(treeRoot, document.Project.Solution.Workspace);
```
Comment (mavasani) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferStringContainsOverIndexOf.Fixer.cs:74]: `DocumentEditor` derives from `SyntaxEditor`, you don't need both. For most cases, using `SyntaxEditor` should suffice.

```suggestion
                    var editor = new SyntaxEditor(treeRoot, document.Project.Solution.Workspace);
```

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferStringContainsOverIndexOf.Fixer.cs [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+                    invocationOperation = invocationOperationOperand;
+                    otherOperation = binaryOperation.RightOperand;
+                }
+                else
+                {
+                    invocationOperation = (IInvocationOperation)binaryOperation.RightOperand;
+                    otherOperation = binaryOperation.LeftOperand;
+                }
+
+                var instanceOperation = invocationOperation.Instance;
```
Comment (mavasani) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferStringContainsOverIndexOf.Fixer.cs:61]: Defensive check to bail out for all cases except argument length 1 and 2? These seem to be the only cases handled in `ReplaceBinaryOperationWithContains`.

```suggestion
                switch (invocationOperation.Arguments.Length)
                {
                    case 1:
                    case 2:
                        break;
                    default:
                        return;
                }

                var instanceOperation = invocationOperation.Instance;
```

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/PreferStringContainsOverIndexOfTests.cs [hunk 5]
```diff
@@ -728,12 +772,25 @@ private void TestMethod(string str, System.StringComparison comparison)
...
     } 
+}";
+            string csFix = @"
+namespace TestNamespace 
+{ 
+    class TestClass 
+    { 
+        private void TestMethod(string str, System.StringComparison comparison) 
+        { 
+            if (" + notString + @"str.Contains(""This"", comparison)" + @")
```
Comment (mavasani) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/PreferStringContainsOverIndexOfTests.cs:783]: Can we add tests with named arguments with different combination/ordering of arguments in input:
1. `if ([|str.IndexOf(value: ""This"", comparison)" + operatorKind + value + @"|])`
2. `if ([|str.IndexOf(""This"", comparisonType: comparison)" + operatorKind + value + @"|])`
3. `if ([|str.IndexOf(value: ""This"", comparisonType: comparison)" + operatorKind + value + @"|])`
4. `if ([|str.IndexOf(comparisonType: comparison, value: ""This"")" + operatorKind + value + @"|])`

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferStringContainsOverIndexOf.Fixer.cs [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+                            }
+                            else
+                            {
+                                containsInvocation = generator.InvocationExpression(containsExpression, indexOfMethodArguments[0].Syntax, indexOfMethodArguments[1].Syntax);
+                            }
+                        }
+                        else
+                        {
+                            containsInvocation = generator.InvocationExpression(containsExpression, indexOfMethodArguments[0].Syntax, indexOfMethodArguments[1].Syntax);
+                        }
```
Comment (mavasani) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferStringContainsOverIndexOf.Fixer.cs:110]: Will this handle named arguments case when arguments are in reverse order?

```csharp
if ([|str.IndexOf(comparisonType: StringComparison.Ordinal, value: ""This"")" + operatorKind + value + @"|])
```

```suggestion
                        int stringArgumentIndex, ordinalArgumentIndex;
                        if (indexOfMethodArguments[0].Value.Type.SpecialType == SpecialType.System_String)
                        {
                            stringArgumentIndex = 0;
                            ordinalArgumentIndex = 1;
                        }
                        else
                        {
                            stringArgumentIndex = 1;
                            ordinalArgumentIndex = 0;
                        }
                        
                        var ordinalArgumentValue = indexOfMethodArguments[ordinalArgumentIndex].Value;
                        if (ordinalArgumentValue.ConstantValue.HasValue &&
                            ordinalArgumentValue.ConstantValue.Value is int intValue &&
                            (StringComparison)intValue == StringComparison.Ordinal)
                        {
                            containsInvocation = generator.InvocationExpression(containsExpression, indexOfMethodArguments[stringArgumentIndex].Syntax);
                        }
                        else
                        {
                            containsInvocation = generator.InvocationExpression(containsExpression, indexOfMethodArguments[0].Syntax, indexOfMethodArguments[1].Syntax);
                        }
```

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferStringContainsOverIndexOf.Fixer.cs [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+                    }
+                    SyntaxNode newIfCondition = containsInvocation;
```
Comment (mavasani) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferStringContainsOverIndexOf.Fixer.cs:112]:
```suggestion

                    SyntaxNode newIfCondition = (int)otherOperation.ConstantValue.Value != -1 ?
                        containsInvocation :
                        generator.LogicalNotExpression(containsInvocation);
```

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferStringContainsOverIndexOf.Fixer.cs [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+                    int rightValue = (int)otherOperation.ConstantValue.Value;
+                    if (rightValue == -1)
+                    {
+                        newIfCondition = generator.LogicalNotExpression(containsInvocation);
+                    }
+                    newIfCondition = newIfCondition.WithTriviaFrom(binaryOperation.Syntax);
```
Comment (mavasani) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferStringContainsOverIndexOf.Fixer.cs:118]: Nit: newline
```suggestion

                    newIfCondition = newIfCondition.WithTriviaFrom(binaryOperation.Syntax);
```

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferStringContainsOverIndexOf.Fixer.cs [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+                    otherOperation = binaryOperation.LeftOperand;
+                }
+
+                var instanceOperation = invocationOperation.Instance;
+
+                context.RegisterCodeFix(
+                    CodeAction.Create(
+                        title: MicrosoftNetCoreAnalyzersResources.PreferStringContainsOverIndexOfTitle,
+                        createChangedDocument: c => ReplaceBinaryOperationWithContains(doc, instanceOperation.Syntax, root, invocationOperation.Arguments, binaryOperation, c),
+                        equivalenceKey: MicrosoftNetCoreAnalyzersResources.PreferStringContainsOverIndexOfMessage),
```
Comment (mavasani) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferStringContainsOverIndexOf.Fixer.cs:67]:
```suggestion
                        equivalenceKey: MicrosoftNetCoreAnalyzersResources.PreferStringContainsOverIndexOfTitle),
```

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferStringContainsOverIndexOf.Fixer.cs [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+                    DocumentEditor editor = await DocumentEditor.CreateAsync(document, cancellationToken).ConfigureAwait(false);
+                    SyntaxEditor syntaxEditor = new SyntaxEditor(treeRoot, document.Project.Solution.Workspace);
+                    SyntaxGenerator generator = editor.Generator;
+                    var containsExpression = generator.MemberAccessExpression(syntaxNode, "Contains");
+                    SyntaxNode? containsInvocation = null;
+                    int numberOfArguments = indexOfMethodArguments.Length;
+                    if (numberOfArguments == 1)
+                    {
+                        var firstArgument = indexOfMethodArguments[0];
+                        if (firstArgument.Parameter.Type.Equals(charType))
```
Comment (mavasani) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferStringContainsOverIndexOf.Fixer.cs:82]:
```suggestion
                        if (firstArgument.Parameter.Type.SpecialType == SpecialType.System_Char))
```

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferStringContainsOverIndexOf.Fixer.cs [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+            {
+                SemanticModel semanticModel = await doc.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+                var compilation = semanticModel.Compilation;
+                if (!compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemString, out INamedTypeSymbol? stringType) ||
+                    !compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemChar, out INamedTypeSymbol? charType) ||
+                    !compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemStringComparison, out INamedTypeSymbol? stringComparisonType))
+                {
+                    return;
+                }
+
```
Comment (mavasani) [inline src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferStringContainsOverIndexOf.Fixer.cs:40]: Not sure we need these checks:
```suggestion
```

**Comment by mavasani** [comment]
> int index = str.IndexOf("AString");
if (index == -1)

Can we file a follow-up issue to handle the simplest case here where index computation is immediately followed by an index check in the next statement? I think it maybe useful, even if it is implemented only for C#. We can certainly wait for feedback before implementing it.

File: src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferStringContainsOverIndexOf.Fixer.cs (hunk 11 above)
Comment (pgovind) [reply src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferStringContainsOverIndexOf.Fixer.cs:40]: Yup, removed.

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/PreferStringContainsOverIndexOfTests.cs (hunk 5 above)
Comment (pgovind) [reply src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/PreferStringContainsOverIndexOfTests.cs:783]: Done. The only issue was for case 3 in VB. I've commented in the unit test

File: src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/PreferStringContainsOverIndexOfTests.cs [hunk 12]
```diff
@@ -361,15 +366,325 @@ End Sub
...
+        End Sub
+    End Class
+End Class
+";
+            vbFix = @"
+Public Class StringOf
+    Class TestClass
+        Public Sub Main()
+            Dim Str As String = ""This is a statement""
+            If Not Str.Contains(value:= " + startQuote + input + startQuote + vbCharLiteral + @", comparisonType:= System.StringComparison.OrdinalIgnoreCase) Then
```
Comment (pgovind) [inline src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/PreferStringContainsOverIndexOfTests.cs:513]: Note that the arguments are reversed here. I don't understand why this happens, but in the debugger I can see that `invocationOperation.SyntaxNode` (in the fixer code) has its arguments in the `(comparisonType, value)` order, but `invocationOperation.Arguments` returns it as (`value, comparisonType)`. It seems to work fine for C# though

Comment (mavasani) [reply src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/PreferStringContainsOverIndexOfTests.cs:513]: @pgovind Argument order in the IOperation world is in the order of runtime evaluation, does not depend on source ordering, see https://github.com/dotnet/roslyn/blob/c61052a5dda190b19abde1bab0064e1bb4bee0ce/src/Compilers/Core/Portable/Generated/Operations.Generated.cs#L541-L548

Comment (mavasani) [reply src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/PreferStringContainsOverIndexOfTests.cs:513]: Sounds good - feel free to file a tracking issue for VB. We can take a look at it based on customer feedback.

Comment (pgovind) [reply src/NetAnalyzers/UnitTests/Microsoft.NetCore.Analyzers/Runtime/PreferStringContainsOverIndexOfTests.cs:513]: https://github.com/dotnet/roslyn-analyzers/issues/3655

**Comment by pgovind** [comment]
Thanks for the review @mavasani

**Pull request merged**